script/extract_google_forum_content.py

```diff
@@ -0,0 +1,164 @@
+            message['in_reply_to'] = '<a3c6d780-e1ed-4409-bfec-d925813df50c@googlegroups.com>'
+
+        # misordered
+        if message['message_id'] == '<138b584a-28ef-46da-ba35-a913cd620c16@googlegroups.com>':
+            message['date'] = 'Wed, 2 Nov 2017 14:36:37 -0800'
+        if message['message_id'] == '<9d13d7dc-0611-43ca-9a0c-418be8469b5c@googlegroups.com>':
+            message['date'] = 'Wed, 1 Dec 2017 12:34:56 -0700'
+
+        # posted forum message with different email than used in slack
+        if message['from'] == 'scott@stratasan.com':
+            message['from'] = 'scott.s.burns@gmail.com'
+        if message['from'] == 'jnbrymn@github.com':
+            message['from'] = 'jfberryman@gmail.com'
+        if message['from'] == 'rmatsum@g.clemson.edu':
+            message['from'] = 'ray.a.matsumoto@vanderbilt.edu'
+
+    return messages
+
+
+def get_chats(messages):
+    messages_by_id = {}
+    for message in messages:
+        messages_by_id[message['message_id']] = message
+
+    unsorted_chats = {}
+    homeless_messages = []
+    for message in messages:
+        original_message = message
+        while message['in_reply_to'] is not None:
+            try:
+                message = messages_by_id[message['in_reply_to']]
+            except KeyError:
+                homeless_messages.append(message)
+                break
+        unsorted_chats.setdefault(message['message_id'], []).append(original_message)
+
+    assert len(homeless_messages) == 0
+
+    sorted_chats = {}
+    for k, v in unsorted_chats.items():
+        subchats = sorted(v, key=lambda m: m['date'])
+        sorted_chats[k] = subchats
+        assert 'Re:' not in subchats[0]
+
+    unsorted_chat_list = []
+    for k, v in sorted_chats.items():
+        unsorted_chat_list.append({
+            'subject': v[0]['subject'],
+            'from': v[0]['from'],
+            'date': v[0]['date'],
+            'messages': v,
+        })
+
+    return sorted(unsorted_chat_list, key=lambda c: c['date'])
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
+        description='Extract and normalize content from penny university google forum. '
+                    'Requires that you retrieve forum content here: https://takeout.google.com'
+    )
+    parser.add_argument(
+        '--data_dump_path', dest='data_dump_path', action='store', help='unzipped google data dump path', required=True,
+    )
+    parser.add_argument(
+        '--output_file', dest='output_file', action='store', help='where to dump the output', required=True,
+    )
+    args = parser.parse_args()
+
+    mbox_path = os.path.join(args.data_dump_path, 'topics.mbox')
+    print(mbox_path)
+    mbox = mailbox.mbox(mbox_path)
+    messages = get_messages(mbox)
+    messages = special_case(messages)
+    chats = get_chats(messages)
+    with open(args.output_file, 'w') as f:
+        f.write(json.dumps(chats))
```

Comment:
Picky: New line at end of file?

Reply:
This file ends in a new line. Typically the Python linter wants there to be a new line at the end of python files. Help me understand if I got something wrong here. Not seeing it.